src-tauri/src/tools.rs

```diff
@@ -1,22 +1,92 @@
 use crate::globals::{
     get_ip_api_key, get_opencage_key, get_reqwest_client, get_weather_api_user_agent,
 };
+use crate::permissions::{check, Permission, Permission::*};
 use base64::prelude::{Engine as _, BASE64_STANDARD_NO_PAD};
 use chrono::prelude::Local;
 use clipboard::{ClipboardContext, ClipboardProvider};
 use image::{
     codecs::png::PngEncoder, imageops::resize, imageops::FilterType::Triangle, ColorType::Rgba8,
     ImageBuffer, ImageEncoder, Rgba,
 };
+use lazy_static::lazy_static;
 use scrap::{Capturer, Display};
-use serde_json::Value;
+use serde_json::{Map, Value};
 use std::{
-    fs::File, io::ErrorKind::WouldBlock, path::Path, thread::sleep, time::Duration
+    fs::File, future::Future, io::ErrorKind::WouldBlock, path::Path, pin::Pin, sync::Arc, thread::sleep, time::Duration
 };
 use urlencoding::encode;
 
-pub async fn get_location_coordinates(location: &str) -> String {
-    println!("getting {} coordinates!", location);
+type SyncAction = dyn Fn(Map<String, Value>) -> String + Send + Sync;
+type AsyncAction = dyn Fn(Map<String, Value>) -> Pin<Box<dyn Future<Output = String> + Send>> + Send + Sync;
```

Comment:
This looks like black magic lol. Maybe a comment here explain this would be helpful?

Reply:
Tried my best to create a comment with my understanding of how that works. Its a bit of magic for sure

Reply:
Those comments are super helpful, thanks!